### DISCUSSION TITLE
Feature Request: HTTP Exception Handling

### DISCUSSION THREAD

**Discussion by snoodlenay** [elimintz/justpy#425] [Ideas]
I have a interest for custom HTTP Exception pages — specifically 404 and 500. I tried passing exception_handlers to the Starlette() instance, as specified at https://www.starlette.io/exceptions/, but I could not get it to work. I’m guessing you can, but meanwhile, I went and kludged together a workaround. Notes, below, in case anyone finds it useful.

1.  Additional import in justpy.py- necessary for both 404 and 500 
`from starlette.responses import HTMLResponse` 

2. Additional parameters to justpy.justpy() — http_404 and http_500. These should be set to a function that, when called, returns HTML. 
```
def justpy(func=None, *, start_server=True, websockets=True, host=HOST, port=PORT, http_404=None, http_500=None, startup=None, **kwargs):
    global func_to_run, startup_func, HOST, PORT, HTTP_404, HTTP_500
    ...
    HTTP_404 = http_404
    HTTP_500 = http_500
```
    
3. In my app, I define my error-handling functions.
```
def server_error():
    wp = justpy.WebPage()
    justpy.Div(a=wp, text="success! we have failed!")
    return wp.to_html()

def not_found():
    wp = justpy.WebPage()
    justpy.Div(a=wp, text="not a page")
    return wp.to_html()
```
Again, I’ve set this up to pass HTML, so anything that generates HTML — a string, rendering a Jinja template, etc. — could go here.

And I pass them to my justpy instance
```
if __name__ == '__main__':
    justpy.justpy(host=HOST, port=PORT, http_404=not_found, http_500=server_error)
```
Note, I’m not passing a func to justpy(). More on that below.

4. Finally, back in justpy.get(), I add the handlers.

**500 handler**
The block that begins at line 141 and starts with
`if inspect.iscoroutinefunction(func_to_run):`
has been modified as follows
```
if inspect.iscoroutinefunction(func_to_run):
    if func_parameters == 1:
        load_page = await func_to_run(request)
    else:
        load_page = await func_to_run()
else:
    if func_parameters == 1:
        if HTTP_500:
            try:
                load_page = func_to_run(request)
            except:
                return HTMLResponse(content=HTTP_500(), status_code=500)
        else:
            load_page = func_to_run(request)
    else:
        load_page = func_to_run()
```

**404 handler**
After the loop through Route.instances (lines 134-138), this has been added
```
if func == initial_func:
    return HTMLResponse(content=HTTP_404(), status_code=404)
```
Note, that this works by testing for initial_func, defined a bit higher on the page, and therefore only works if NO function is passed to justpy.justpy(). I believe that addressing that case might involve a new approach to matching routes, that doesn’t match ‘/‘ by default, and I do not yet grasp the architecture enough to fiddle with that knob.

**Comment by elimintz** [comment]
The way I was thinking about it was that the 404 handler would be the function provided to justpy as a parameter. All the valid URLs are handled as routes and when no route is matched, the function is run. Let me know what you think.

For the 500 exception, there does need to be a change and I will add that.

**Comment by snoodlenay** [comment]
I think that sounds like the best way to handle the 404. When I followed the trail through justpy.routing.Route to starlette.routing.compile_path(), I got the impression that Starlette falls back to matching everything to '/' ... and I both a) didn't see a quick workaround and b) reckoned I was missing something, because that just seems like a silly rule to have. I figured I'd leave it for you. :)
